### ISSUE TITLE
Is there a recommended structure for Semi-SL with model co-learning?

### ISSUE THREAD

**Issue by dreaquil** [PyTorchLightning/pytorch-lightning#10587]
### Discussed in https://github.com/PyTorchLightning/pytorch-lightning/discussions/10561

<div type='discussions-op-text'>

<sup>Originally posted by **dreaquil** November 16, 2021</sup>
Hi! Recently started using pytorch lightning and really enjoying it so far! However, I'm still getting used to the way of doing things. I'm currently working on implementing  a [DivideMix](https://arxiv.org/pdf/2002.07394.pdf) model which is based on MixMatch (Semi supervised learning). This has a rather complicated training loop as it contains two models and uses the output of one model to sample the training data for the other. I'm struggling to figure out how best to fit this into the pytorch lightning framework or if this may be beyond the framework's structure. Does anyone have any idea what the best way to structure this would be?

Training starts with:
- A warm up period for both models (first 10 epochs or so).

From my understanding, the following steps need to take place on each epoch:

- Loss calculation for each sample in the full training set by both models.
- Fitting a two component GMM to the loss distribution for each model.
- Using the GMM produced by model 2 to split training data into clean and noisy data for training model 1
- Label refinement/ Label Co-guessing for model 1 inputs (Still a bit hazy)
- Mixing inputs
- Optimization for model 1
- Using the GMM produced by model 1 to split training data into clean and noisy data for training model 2
- Label refinement / Label Co-guessing for model 2 inputs (Still a bit hazy)
- Mixing inputs
- Optimization for model 2</div>

**Comment by dreaquil** [comment]
Turned out to be fairly easy to integrate this! Well done to the entire lightning team for such a flexible code base. For anyone looking for advice, the hook system was really useful in implementing dividemix. Additionally, I needed to turn on manual optimization.
